feat: log bandwidth on substream instead of socket level

melekes opened the pull request:
## Description

Previously, the `with_bandwidth_logging` extension to `Transport` would track the bytes sent and received on a socket level. This however only works in conjunction with `Transport` upgrades where a separate multiplexer is negotiated on top of a regular stream.

With QUIC and WebRTC landing, this no longer works as those `Transport`s bring their own multiplexing stack. To still allow for tracking of bandwidth, we refactor the `with_bandwidth_logging` extension to count the bytes send on all substreams opened through a `StreamMuxer`. This works, regardless of the underlying transport technology. It does omit certain layers. However, there isn't necessarily a "correct" layer to count bandwidth on because you can always go down another layer (IP, Ethernet, etc).

Closes #3157.

## Change checklist

<!-- Please add a Changelog entry in the appropriate crates and bump the crate versions if needed. See <https://github.com/libp2p/rust-libp2p/blob/master/docs/release.md#development-between-releases>-->

- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] A changelog entry has been made in the appropriate crates

melekes committed:
log bandwidth at `StreamMuxer` level

An alternative approach to #3161 suggested in https://github.com/libp2p/rust-libp2p/issues/3157#issuecomment-1326126621

Closes #3157

thomaseizinger commented on swarm/src/lib.rs:
```diff
@@ -457,6 +459,24 @@ where
+        })
+        .boxed();
+        sinks
+    }
```
I don't think this should live at the `Swarm` level. Instead, I think we can have this in `libp2p` same as before but implement `TransportExt` for every `Transport` that outputs `(PeerId, S)` with `S: StreamMuxer`. Does that compile?

melekes replied:
something like this: https://github.com/libp2p/rust-libp2p/pull/3180/commits/fac29fc95ed100f5fd7efedca8a1ac57e752dd83 ?

thomaseizinger replied:
Yep, thanks!

melekes committed:
new transport ext

melekes committed:
do not break the public API

melekes committed:
rename TInner to SMInner

melekes committed:
rename connections to streams

melekes committed:
add changelog entry

melekes committed:
bring back `BandwidthFuture`

to avoid breaking API

melekes committed:
make `new_with_sinks` and `BandwidthSinks::new` pub

only within this crate

melekes committed:
revert changes to transport_ext visibility

mxinden commented on CHANGELOG.md:
```diff
@@ -45,6 +45,10 @@
 
 # `libp2p` facade crate
 
+# Unreleased
```
```suggestion
# 0.51.0 [unreleased]
```

mxinden commented on CHANGELOG.md:
```diff
@@ -45,6 +45,10 @@
 
+# Unreleased
+
+- Count bandwidth at the application level ([PR 3180](https://github.com/libp2p/rust-libp2p/pull/3180)).
```
```suggestion
- Count bandwidth at the application level. Previously `BandwidthLogging` would implement `Transport` and now implements `StreamMuxer` ([PR 3180](https://github.com/libp2p/rust-libp2p/pull/3180)).
```

mxinden commented on src/bandwidth.rs:
```diff
@@ -43,120 +36,98 @@ use std::{
 
-/// Wraps around a `Transport` and counts the number of bytes that go through all the opened
-/// connections.
+/// Wraps around a `StreamMuxer` and counts the number of bytes that go through all the opened
```
```suggestion
/// Wraps around a [`StreamMuxer`] and counts the number of bytes that go through all the opened
```

mxinden commented on src/bandwidth.rs:
```diff
@@ -177,13 +148,13 @@ impl BandwidthSinks {
 /// Wraps around an `AsyncRead + AsyncWrite` and logs the bandwidth that goes through it.
 #[pin_project::pin_project]
-pub struct BandwidthConnecLogging<TInner> {
+pub struct BandwidthConnecLogging<SMInner> {
```
```suggestion
pub struct InstrumentedStream<SMInner> {
```

This is no longer a connection, thus I would suggest a rename. How about `InstrumentedStream`?

mxinden commented on src/bandwidth.rs:
```diff
@@ -177,13 +148,13 @@ impl BandwidthSinks {
 
 /// Wraps around an `AsyncRead + AsyncWrite` and logs the bandwidth that goes through it.
```
```suggestion
/// Wraps around an [`AsyncRead`] + [`AsyncWrite`] and logs the bandwidth that goes through it.
```

While we are at it, if you don't mind.

mxinden commented on src/bandwidth.rs:
```diff
@@ -252,3 +223,13 @@ impl<TInner: AsyncWrite> AsyncWrite for BandwidthConnecLogging<TInner> {
+/// counter.
+#[deprecated(since = "0.50.1")]
+#[pin_project::pin_project]
+pub struct BandwidthFuture<TInner> {
```
Changing `BandwidthLogging` to implement `StreamMuxer` instead of `Transport` is a breaking change. Given that you are proposing a breaking change anyways, why deprecate here?

melekes replied:
Oh, right. Will remove

mxinden reviewed (commented):
Direction looks good to me. A couple of comments.

melekes committed:
amend changelog and comments

Co-authored-by: Max Inden <mail@max-inden.de>

melekes committed:
remove BandwidthFuture

also
rename BandwidthLogging `new_with_sinks` to `new`
rename BandwidthConnecLogging to InstrumentedStream

melekes committed:
Merge branch 'master' into anton/3157-bandwidth-logging-streammuxer

mxinden commented on src/transport_ext.rs:
```diff
@@ -31,11 +39,31 @@ pub trait TransportExt: Transport {
     ///
```
Comment above needs updating. We are no longer logging at the socket layer.

mxinden commented on src/transport_ext.rs:
```diff
@@ -31,11 +39,31 @@ pub trait TransportExt: Transport {
     ///
     /// This method returns an `Arc<BandwidthSinks>` that can be used to retreive the total number
```
```suggestion
    /// This method returns an `Arc<BandwidthSinks>` that can be used to retrieve the total number
```

mxinden commented on src/transport_ext.rs:
```diff
@@ -31,11 +39,31 @@ pub trait TransportExt: Transport {
     /// This method returns an `Arc<BandwidthSinks>` that can be used to retreive the total number
     /// of bytes transferred through the sockets.
-    fn with_bandwidth_logging(self) -> (BandwidthLogging<Self>, Arc<BandwidthSinks>)
+    fn with_bandwidth_logging<S>(self) -> (Boxed<(PeerId, StreamMuxerBox)>, Arc<BandwidthSinks>)
```
Can you either add a small test, or add a doc example to the `with_bandwidth_logging` method (preferred)?

melekes replied:
Added a doc test (ignored because it requires certain features)

mxinden replied:
Thanks for adding the test.

In my eyes a doc test that is `ignore`d is not helpful and destined to be outdated. Is there some way to enable the features for testing only? //CC @thomaseizinger

thomaseizinger replied:
Doc tests have access to all `dev-dependencies` so you can simply add whatever you need there and it will be available in the test. You will have to change the imports to though.

thomaseizinger commented on src/transport_ext.rs:
```diff
@@ -31,11 +39,31 @@ pub trait TransportExt: Transport {
+            let (peer_id, stream_muxer_box) = output.into();
+            (
+                peer_id,
+                StreamMuxerBox::new(BandwidthLogging::new(
```
If we use `StreamMuxerBox` here, then lets make `BandwidthLogging` and `InstrumentStream` `pub(crate)`!

thomaseizinger reviewed (approved):
Nice work, thank you!

One suggestion, otherwise happy with this modulo @mxinden's comments!

melekes committed:
Merge branch 'master' into anton/3157-bandwidth-logging-streammuxer

melekes committed:
add a doc test

melekes committed:
enable doc test

melekes committed:
Merge branch 'master' into anton/3157-bandwidth-logging-streammuxer

melekes committed:
fix doc build

melekes committed:
Merge branch 'master' into anton/3157-bandwidth-logging-streammuxer

thomaseizinger commented on Cargo.toml:
```diff
@@ -136,6 +136,10 @@ env_logger = "0.10.0"
 clap = { version = "4.0.13", features = ["derive"] }
 tokio = { version = "1.15", features = ["io-util", "io-std", "macros", "rt", "rt-multi-thread"] }
 
```
Why split these with a newline?

thomaseizinger reviewed (dismissed):
LGTM, thanks!

thomaseizinger committed:
Merge branch 'master' into anton/3157-bandwidth-logging-streammuxer

mxinden commented on Cargo.toml:
```diff
@@ -136,6 +136,10 @@ env_logger = "0.10.0"
 clap = { version = "4.0.13", features = ["derive"] }
 tokio = { version = "1.15", features = ["io-util", "io-std", "macros", "rt", "rt-multi-thread"] }
 
+libp2p-mplex = { version = "0.38.0", path = "muxers/mplex" }
+libp2p-noise = { version = "0.41.0", path = "transports/noise" }
+libp2p-tcp = { version = "0.38.0", path = "transports/tcp", features = ["tokio"] }
```
Is it a deliberate choice to link these with a `version`?

mxinden replied:
Would be fixed with https://github.com/libp2p/rust-libp2p/pull/3261.
